### PR TITLE
Add tox ``lint`` target

### LINKED ISSUE #11599: tox targets for non-tests
> @AA-Turner Could I ask what was the rationale for this change? How are contributors expected to run e.g. linters on their changes without the tox targets?

_Originally posted by @stephenfin in https://github.com/sphinx-doc/sphinx/commit/ff852bc7c31f48e66100e4647749fc199d92ca79#commitcomment-124548969_

### PR DESCRIPTION
**Pull request by stephenfin**
### Feature or Bugfix

Refactoring

### Purpose

Re-add a tox environment for linters. This environment uses the 'lint' extras
and runs the same commands as the GitHub Workflows. A note is added reminding
people to keep the two in-sync.

### Detail

This resolves #11599. There is an (arguably superior but more involved) alternative at #11603.

### Relates

(none)

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -26,8 +25,20 @@ setenv =
...
+    Run linters.
+extras =
+    lint
+# If you update any of these commands, don't forget to update the equivalent
+# GitHub Workflow step
+commands =
+    ruff . --diff --format github
+    flake8 .
+    isort --check-only --diff .
+    mypy sphinx/
```
Comment (picnixz) [inline tox.ini:42]: Don't we also apply mypy to tests?

File: tox.ini [hunk 2]
```diff
@@ -1,7 +1,6 @@
 [tox]
-minversion = 2.4.0
+minversion = 4.2.0
 envlist = py{39,310,311,312}
-isolated_build = True
```
Comment (picnixz) [inline tox.ini:4]: Why did you remove the isolated build? (I don't really use tox personally and I run the tests from CLI using specific Python version)

Comment (stephenfin) [reply tox.ini:4]: All the context is in the commit messages. From [`f2f2e79` (#11602)](https://github.com/sphinx-doc/sphinx/pull/11602/commits/f2f2e79a9d5a221c50f3eeb08dc0dfce48ac5d08):

> `isolated_build` is no longer required since builds are isolated by default in tox 4.
> 
> `basepython` is no longer required since tox 4 only supports Python 3 and 'basepython' defaults to the version of Python that tox is installed with.

Comment (picnixz) [reply tox.ini:4]: Oh sorry, I didn't see the commit message itself. My bad !

File: tox.ini (hunk 1 above)
Comment (stephenfin) [reply tox.ini:42]: Apparently not: I simply copied these commands from `.github/workflows/lint.yml`

Comment (picnixz) [reply tox.ini:42]: Oh good to know. @AA-Turner any reason why we are not applying mypy to tests actually ? (is it perhaps because there would be too many false positive due to monkey patching or other weird constructions?)

Comment (stephenfin) [reply tox.ini:42]: > (is it perhaps because there would be too many false positive due to monkey patching or other weird constructions?)

Not @AA-Turner but that's usually why I don't run mypy over tests, yes.

**Comment by electric-coder** [comment]
@picnixz 

> Oh good to know. @AA-Turner any reason why we are not applying mypy to tests actually? 

Most likely because it can be a lot of extra work and maintainers didn't prioritize it. However, it would improve code quality.

@stephenfin

> Not @AA-Turner but that's usually why I don't run mypy over tests, yes.

Absolutely worth it even if you have to use `# ignore` on occasion.
